specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,378 @@
+* The app will query ApplifeCycle APIs to retrieve an ActivationKind. (Note: This is covered in a seperate AppLifeCycle API spec.)
+* If the Activation is Push, the app will QI the Arguments property to retrieve an instance of PushActivatedEventArgs and get the Push payload from it. Care needs to be taken to Take a Deferral and Release the Deferral while processing the payload for Power Management compliance.
+* If the Activation is Foreground, the app will do 2 things:
+  * It will subscribe to a Foreground In-memory Push event
+  * It will request a Push Channel Asynchronously with an implementation of Progress and Completed event handler.
+    * Expect Channel operations to take around 2-16 minutes in some rare cases (retryable errors). In the 90th percentile case, it should be fairly quick operation (in a few seconds).  
+
+```cpp
+int main()
+{
+    // Always Register the Push trigger as the first step: It sets up the Inproc COM server for subsequent Push operations
+    winrt::guid activatorGuid{ 0xC380465D, 0x2271, 0x428C, { 0x9B, 0x83, 0xEC, 0xEA, 0x3B, 0x4A, 0x85, 0xC1} };
+    InProcActivatorDetails details(activatorGuid);
+    PushManager::RegisterActivator(details);
+
+    // Check to see if the WinMain activation is due to a Push Activator
+    // Note: This API is currently not present in Reunion but will be included as part of the AppLifecycle Public API spec.
+    auto activation = AppLifecycle::Activation();
+    auto kind = activation.Kind();
+
+    if (kind == ReunionActivationKind::Push)
+    {
+        auto pushArgs = activation.Arguments().try_as<PushActivatedEventArgs>();
+
+        // Call TakeDeferral to ensure that code runs in low power
+        pushArgs.GetDeferral();
+
+        auto payload = pushArgs.Payload();
+
+        // Do stuff to process the raw payload
+
+        // Call CompleteDeferral as good practise: Needed mainly for low power usage
+        pushArgs.CompleteDeferral();
+    }
+    else if (kind == ReunionActivationKind::Launch) // This indicates that the app is launching in the foreground
+    {
+        // Register an event to Intercept Push payloads
+        auto eventToken = details.PushActivated([](const auto&, PushActivatedEventArgs args)
+        {
+                // Call TakeDeferral to ensure that code runs in low power
+                args.GetDeferral();
+
+                auto payload = args.Payload();
+
+                // Do stuff to process the raw payload
+
+                // Call CompleteDeferral as good practise: Needed mainly for low power usage
+                args.CompleteDeferral();
+        });
+
+        // {F80E541E-3606-48FB-935E-118A3C5F41F4}
+        winrt::guid remoteId{ 0xf80e541e, 0x3606, 0x48fb, {0x93, 0x5e, 0x11, 0x8a, 0x3c, 0x5f, 0x41, 0xf4} };
+        auto channelOperation = PushManager::CreateChannelAsync(remoteId);
```

Comment:
@andrewleader Would you feel as strongly if c++/winrt offered a make_guid function that takes in a string? 

I believe it was added to c++winrt around dec 31 2020:
https://github.com/microsoft/cppwinrt/issues/797

Reply:
I don't think that'd change my opinion. C# also has `new Guid("F80E541E-3606-48FB-935E-118A3C5F41F4")`, yet both of those aforementioned SDKs chose to take a string instead of a GUID. It seems like existing precidence for these is to take a string, I think we'd need some strong justification to change that?

Reply:
Gotcha. This was actually a suggestion by @riverar that I made the change. If string is the pattern followed, I can go back to it.

@sharath2727 @danielayala94 FYI

Reply:
Ah cool, @riverar do you have examples of other Microsoft APIs (or other industry APIs) which take GUIDs for their identifiers instead of strings like the two I showed? Or reasoning why moving forward, new APIs should take GUIDs instead of strings?

Reply:
@andrewleader I was thinking about the many forms a GUID string can take. With braces. Without braces. With hyphens. Without hyphens. Set of raw hexadecimal letters. Using a structure creates a canonical form that I believe reduces potential incident rate.

winrt::make_guid seems like a decent fix to the raw guid mess and aligns with C# development (new Guid("...")).

I drew inspiration from some APIs in WinRT:
* IMetaDataDispenser::DefineScope(REFCLSID rclsid, ...)
* IBackgroundTaskBuilder5::SetTaskEntryPointClsid(GUID TaskEntryPoint)
* IBackgroundTaskCompletedEventArgs::InstanceId(GUID* value)
* IDesignerAppManager::LoadObjectIntoAppAsync(..., GUID classId, ...)
* IDeviceAccessInformationStatics::CreateFromDeviceClassId(GUID deviceClassId, ...)

But I acknowledge this isn't consistent across the board:
* IGetActivationFactory::GetActivationFactory(HSTRING activatableClassId, ...)
* IAudioDeviceModule::ClassId(HSTRING* value)

But perhaps the _real_ fix is to eliminate the parameter altogether and create it/cache it internally?

Reply:
Regarding the C++ example above, you write it as follows:

```C++
PushManager::CreateChannelAsync(guid("F80E541E-3606-48FB-935E-118A3C5F41F4"));
```

If the identifier _must_ be a `guid` then the API should expect a `guid`. If the identifier can be something other than a `guid` then a string makes sense. If you can get away with not asking the developer for an identifier that would be the best. 

Expecting a `guid` in a string format just adds a tax. The implementation must ensure that the `guid` is properly formatted. The developer has to read the docs to figure out that the string is actually a `guid`. Odds are they won't and you've replaced an IntelliSense/compiler error (good) for a runtime error (bad). But I don't know whether the `guid` is a requirement or not.

> codified in a style guide

I think @jonwis has such a thing, no?

Reply:
@hulumane if we are going to support bring your own identity in the future then probably this should be a string?

Reply:
I haven't heard of any plans on supporting "Bring your own identity" with RemoteId Reunion APIs. When we do have a plan for that, we should simply introduce an API overload that takes in a String, IMHO. @nicuparente did we ever talk about reunion supporting BYOI?

Reply:
@hulumane BYOI = BringYourOwnIdentity? I haven't heard that phrase before. Could you elaborate?

Reply:
@DrusTheAxe 
It was a discussion with our server counterparts last year where we contemplated that apps like Contoso could bring their own remote identifier (orthogonal to AAD) for authentication with WNS. For the Windows Platform which uses a single socket for all apps on the device, the POR is to use RemoteIdentifier = AAD GUID and the BYOI conversation isn't relevant IMHO.